src/Symfony/Component/FeatureToggle/Strategy/DateStrategy.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\FeatureToggle\Strategy;
+
+use Symfony\Component\FeatureToggle\StrategyResult;
+use Psr\Clock\ClockInterface;
+
+final class DateStrategy implements StrategyInterface
+{
+    public function __construct(
+        private readonly ClockInterface $clock,
+        private readonly \DateTimeImmutable|null $from = null,
+        private readonly \DateTimeImmutable|null $until = null,
+        private readonly bool $includeFrom = true,
+        private readonly bool $includeUntil = true,
+    ) {
+        if (null === $this->from && null === $this->until) {
+            throw new \InvalidArgumentException('Either from or until must be provided.');
+        }
+    }
+
+    public function compute(): StrategyResult
+    {
+        $now = $this->clock->now();
+
+        if (null !== $this->from) {
+            if ($this->includeFrom && $this->from > $now) {
+                return StrategyResult::Deny;
+            }
+
+            if (!$this->includeFrom && $this->from >= $now) {
+                return StrategyResult::Deny;
+            }
+        }
+
+        if (null !== $this->until) {
+            if ($this->includeUntil && $this->until < $now) {
+                return StrategyResult::Deny;
+            }
+
+            if (!$this->includeUntil && $this->until <= $now) {
+                return StrategyResult::Deny;
+            }
+        }
+
+        return StrategyResult::Grant;
```

Comment:
should this grant or abstain ?

Reply:
I didn't find any valid use case for Abstain in this strategy. You are either valid according to dates or not. There is "I can't answer" Maybe the Deny part could be kind of dynamic to allow either Deny or Abstain ? But I think that would make things much more complicated to understand.

Reply:
Well, with those strategies that return a result based on the first one that does not abstain, it is important to define clearly those cases to make combinators useful.

However, with the current API of the component where each Feature holds a strategy, the abstain case looks quite useless. You would wire the strategy you need, skipping the irrelevant one entirely (or having a default as part of the config of that strategy alone).

For security voters, the abstain state is very important because the authorization checker does not have different voters for each permission being checked. It has a single list of voters (and a strategy about how their decisions should be combined).

Reply:
We allow custom strategies and nested strategies. Abstain allows to fallback to other strategies easily. We could get rid of it with syntax like ((a OR b) AND c) but the problem is then how do we parse this to automatically create the strategies (with the right configuration)

Reply:
Custom strategies and nested strategies don't need a feature to abstain. The possibility to abstain only makes those cases harder.
And as each feature currently has its own strategy object, I see to reason to register strategies that will abstain. Registering the exact strategy it needs is much easier to understand.

A case like `(a OR b) AND c` would be instantiated as `new AndStrategy([new OrStrategy([new A(), new ()]), new C()])`. And `AndStrategy` and `OrStrategy` are much simpler to implement if they don't need to care about potential abstain results.

With your current combined strategies, it is not even possible to represent that case (AffirmativeStrategy is a OrStrategy as any grant provides a grant, but there is no AndStrategy). And PriorityStrategy is meaningful only because of abstain, while I see no case where I would actually want this behavior for an understandable behavior (especially when many strategies never abstain).

> We allow custom strategies and nested strategies. Abstain allows to fallback to other strategies easily. We could get rid of it with syntax like ((a OR b) AND c) but the problem is then how do we parse this to automatically create the strategies (with the right configuration)

to me, we should not have both an expression syntax with functions corresponding to some strategies and operators while having a single strategies where combining other strategies together can be done in custom ways. those are simply 2 different models which will be hard to reconcile (your PriorityStrategy has no sane representation as an expression for instance)

Reply:
It is not simply cases or OR and AND. More like `a == grant || b == grant` , `a != abstain || b != abstain`, ... Abstain makes teh difference between "I couldn't answer ask someone else or use default" and "I was able to answer and the answer is to not activate it".

`PriorityStrategy` is represented like this : `a != abstain || b != abstain`. That is waht I meant by "parsing" might be hard is we need to check for the operator + the value it is compared to.

`Affirmative` is `a == grant || b == grant`

`Consensus` would be `a != abstain && b != abstain` and so on.